### PR TITLE
fix(extrinsic): ensure signature includes key 

### LINKED ISSUE #620: Security Issues with add_public_key_to_msa
In review, `add_public_key_to_msa` was found to have two issues:
- [ ] The signature of the `new_key_owner_proof` was not being registered against replay attacks
- [ ] `add_key_payload` did not include the new key owner, allowing the `msa_owner_proof` to be swapped to a different `new_public_key` by the party submitting the transaction

### PR DESCRIPTION
**Pull request by enddynayn**
# Goal
Ensure the AddKeyData struct includes the new
public-key being added. Including the public-key
in the signature prevents a provider from
being able to add any key.

Additionally, register the proof for the new
public-key to prevent replay attacks.

Update benchmarks for add-public-key-to-msa
and delete-msa-public-key.

Closes #620

### CODE REVIEW

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -562,37 +562,43 @@ pub mod pallet {
 		#[pallet::weight(T::WeightInfo::add_public_key_to_msa())]
 		pub fn add_public_key_to_msa(
 			origin: OriginFor<T>,
-			msa_owner_public_key: T::AccountId,
+			delegator_key: T::AccountId,
```
Comment (wilwade) [inline pallets/msa/src/lib.rs:565]: I don't think we want to confuse adding keys with delegating. I would suggest keeping the original name.

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -562,37 +562,43 @@ pub mod pallet {
...
-				.map_err(|_| Error::<T>::AddKeySignatureVerificationFailed)?;
-
+			Self::register_signature(&msa_owner_proof, add_key_payload.expiration.into())?;
 			Self::register_signature(&new_key_owner_proof, add_key_payload.expiration.into())?;
 
-			let msa_id = add_key_payload.msa_id;
+			let delegator_msa_id = add_key_payload.msa_id;
 
-			Self::ensure_msa_owner(&msa_owner_public_key, msa_id)?;
+			Self::ensure_msa_owner(&delegator_key, delegator_msa_id)?;
```
Comment (wilwade) [inline pallets/msa/src/lib.rs:589]: Same here. I think using delegator terminology here is confusing.

File: pallets/msa/src/types.rs [hunk 3]
```diff
@@ -17,12 +17,15 @@ use scale_info::TypeInfo;
...
+#[derive(TypeInfo, RuntimeDebugNoBound, Clone, Decode, Encode, PartialEq, Eq)]
+#[scale_info(skip_type_params(T))]
+pub struct AddKeyData<T: Config> {
 	/// Message Source Account identifier
 	pub msa_id: MessageSourceId,
 	/// The block number at which the signed proof for add_public_key_to_msa expires.
-	pub expiration: BlockNumber,
+	pub expiration: T::BlockNumber,
+	/// The public key to be added.
+	pub new_public_key: T::AccountId,
```
Comment (wilwade) [inline pallets/msa/src/types.rs:28]: Could we skip the `T` by using `AccountId32` here instead?

Comment (enddynayn) [reply pallets/msa/src/types.rs:28]: how about we pass the generic params instead like so:

```
pub struct AddKeyData<BlockNumber, AccountId> {...}
```
This allows us to be consistent with grabbing the type from system.

**Review by wilwade — APPROVED** [review]
💯

File: pallets/messages/src/lib.rs [hunk 4]
```diff
@@ -224,6 +224,7 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::add_ipfs_message(cid.len() as u32, message.index as u32)).into())
 		}
+
```
Comment (shannonwells) [inline pallets/messages/src/lib.rs:227]: if there are no other changes, maybe revert this file

File: pallets/msa/src/benchmarking.rs [hunk 5]
```diff
@@ -65,6 +74,15 @@ fn create_account_with_msa_id<T: Config>(n: u32) -> (T::AccountId, MessageSource
 	(provider.clone(), msa_id)
 }
 
+fn create_msa_account_and_keys<T: Config>() -> (T::AccountId, SignerId, MessageSourceId) {
```
Comment (shannonwells) [inline pallets/msa/src/benchmarking.rs:77]: ❤️

File: pallets/msa/src/replay_tests.rs [hunk 6]
```diff
@@ -35,17 +35,17 @@ pub fn user_adds_key_to_msa(
 	delegator_pair: sp_core::sr25519::Pair,
 	new_pair: sp_core::sr25519::Pair,
 ) {
```
Comment (shannonwells) [inline pallets/msa/src/replay_tests.rs:37]: I thought there was a new replay test for the other signature, does there need to be one?

**Review by shannonwells — APPROVED** [review]
Just comments/questions, assuming there is adequate replay test coverage, nothing blocking

File: pallets/msa/src/replay_tests.rs (hunk 6 above)
Comment (enddynayn) [reply pallets/msa/src/replay_tests.rs:37]: You mean this: https://github.com/LibertyDSNP/frequency/pull/617/files#diff-376fd47e95f7a80f67ffe3cf86e50bfd6d4d4d04dbdf54ed76c5a4b06b9c5b36R1609

**Pull request merged**